### ISSUE TITLE
Other Spec Review: <meta name="text-scale" content="scale" />

### ISSUE THREAD

**Issue by davidsgrogan** [w3ctag/design-reviews#1172]
### Specification

https://drafts.csswg.org/css-fonts-5/#text-scale-meta

### Explainer

https://github.com/w3c/csswg-drafts/blob/main/css-env-1/explainers/meta-text-scale.md

### Links

- An introduction to the feature, aimed at unfamiliar audiences: https://github.com/w3c/csswg-drafts/blob/main/css-env-1/explainers/meta-text-scale.md#introduction
- A description of the problems that end-users were facing before this proposal: https://github.com/w3c/csswg-drafts/blob/main/css-env-1/explainers/meta-text-scale.md#user-facing-problem
- Alternatives considered: https://github.com/w3c/csswg-drafts/blob/main/css-env-1/explainers/meta-text-scale.md#alternatives-considered . Also, *most* developer use cases would be solved by using `env(preferred-text-scale)` in CSS calc expressions, but we heard from web developers that they "just want `rem` to work like they were promised" -- i.e. `rem` will honor the users' a11y preferences, no matter if those prefs are specified at the OS-level *or* UA-level. Authors don't want to mess around with `env()`. This `<meta>` proposal allows for that.
- Examples of how to use the proposal to solve the end-users' problems: https://davidsgrogan.github.io/css-fonts-5/Overview.html#ex-scaled-page
- What do the end-users experience with this proposal: Font sizes will honor the users' OS-level a11y preference. See https://github.com/w3c/csswg-drafts/blob/main/css-env-1/explainers/env-preferred-text-scale.md#example-1-reddit part [4] for an illustration of the user-facing benefits
- Web Platform Tests:  None yet

### The specification

- [x] Follows the [Web Platform Design Principles](https://www.w3.org/TR/design-principles/).
- [x] Includes Security and Privacy Considerations sections based on answers to the [Security/Privacy Questionnaire](https://www.w3.org/TR/security-privacy-questionnaire/).

### Where and by whom is the work is being done?

- GitHub repo: https://github.com/w3c/csswg-drafts
- Primary contacts:
  - @davidsgrogan Google, CSSWG member, Spec author, Chrome implementer
  - @joshtumath BBC, CSSWG member, Spec reviewer, Explainer author, Feature champion, Dx pollster
- Organization/project driving the specification: CSSWG, Google, BBC
- This work is being funded by: Google and BBC
- Primary standards group developing this feature: CSSWG
- Incubation and standards groups that have discussed the design:
  - https://github.com/w3c/csswg-drafts/issues/12380#issuecomment-3053293717

### Feedback so far

- Multi-stakeholder feedback:
  - Mozilla comments: https://github.com/mozilla/standards-positions/issues/1326
  - WebKit comments: https://github.com/WebKit/standards-positions/issues/587
  - {{...include feedback/review from developers, implementers, civil society, and others}} -- we've gotten informal positive feedback from web developers at a few conferences (thanks @joshtumath for polling). Other implementers: Apple and Mozilla engineers had vaguely positive comments during the CSS telecon where this was resolved to add, FWIW ([minutes](https://github.com/w3c/csswg-drafts/issues/12380#issuecomment-3053293717)).
- Major unresolved issues with or opposition to this specification: ~~https://github.com/w3c/csswg-drafts/issues/12475 -- the unresolved issue is if we should specify that UAs should scale fonts non-linearly... meaning the larger the font already is, the less of an enlargement multiplier it would get. We [anticipate](https://github.com/w3c/csswg-drafts/issues/12380#issuecomment-3059624945) that shipping _linear_ scaling will NOT prevent from future experiments with _non-linear_ scaling.~~ **(Update: Resolved)** We are not aware of any opposition to the specification.
- Status/issue trackers for implementations:
https://chromestatus.com/feature/5112244702674944
https://bugzilla.mozilla.org/show_bug.cgi?id=1976522

---

Track conversations at https://tag-github-bot.w3.org/gh/w3ctag/design-reviews/1172

**Comment by matatk** [comment]
Hi @davidsgrogan, and thanks for your review request.

Speaking personally, as someone with a vision impairment, this is something I have wanted to see on the web for some time; thank you for working on it!

We have some feedback on the feature, the explainer, and concerns and suggestions about adoption.

Regarding the feature:

* Overall, this seems like a great improvement.

* Have you considered keeping _only_ `legacy` and `scale` as values, and having the UA (on behalf of, and influenced by, the user, the device, and other factors) decide on what sort of scaling (linear or otherwise) to apply? This seems to be a decision on which the user would know best (possibly also guided by their UA and device).

  As you say in the explainer, the reason that `text-scale` has to be opt-in is potential breakage of _existing_ sites. However, new sites that are designed with a more adaptive approach to text sizing could be much more robust against whether the scale is linear or otherwise.

  Have you considered explicitly making this choice up to the UA?

* Regarding the [question as to whether the font size keywords should also scale](https://github.com/w3c/csswg-drafts/issues/12475): we think this makes sense and should be included.

* We'd agree in expecting it to be very likely there would be significant breakage of existing site layouts if scaling were the default. But out of curiosity: have you done any testing to check on this?

Regarding the explainer:

* It's a well-written document, but it would be really helpful to have more illustrative examples - at least one "before and after" style example - of the effect of the proposal.

* Our read on the proposal is that it _would_ affect the base font size on desktop too (i.e. as well as mobile) - but [the table](https://github.com/w3c/csswg-drafts/blob/main/css-env-1/explainers/meta-text-scale.md#comparison-of-legacy-and-scale) makes this seem a little unclear - could you clarify?

Concerns regarding adoption:

From experience working with developers, the huge barrier this has to get over is that people need to opt in. It may be very hard to persuade busy development teams to opt in without some substantial resources to help them get started with such layouts. These may include:

- Guidance for authors in the spec.

- [WAI Tutorials](https://www.w3.org/WAI/tutorials/) for authors to make more adaptive/robust/accessible layouts.

- Making PRs on established projects such as

  + React Starter

  + Svelte

  + anything to do with bootstrapping projects

Getting this into the default `<head>` for projects will help achieve adoption - but it will also need some helpful resources for busy developers to ensure that the DX is still good.

You might want to contact [Docs CG](https://www.w3.org/community/docs-cg/) to see if they can assist you with writing the developer documentation.

We look forward to hearing your thoughts on this.

Thanks to @AutoSponge in APA WG for the benefit of his experience managing development.

**Comment by davidsgrogan** [comment]
Thanks for the great detailed review! Responses inline:

> Hi [@davidsgrogan](https://github.com/davidsgrogan), and thanks for your review request.
> 
> Speaking personally, as someone with a vision impairment, this is something I have wanted to see on the web for some time; thank you for working on it!
> 
> We have some feedback on the feature, the explainer, and concerns and suggestions about adoption.
> 
> Regarding the feature:
> 
> * Overall, this seems like a great improvement.
> * Have you considered keeping _only_ `legacy` and `scale` as values, and having the UA (on behalf of, and influenced by, the user, the device, and other factors) decide on what sort of scaling (linear or otherwise) to apply? This seems to be a decision on which the user would know best (possibly also guided by their UA and device).

Yes! The WG resolution and spec text about linear vs non-linear scaling does not require that a UA always do non-linear scaling (the spec draft has it as a SHOULD -- "user agents should scale these other keywords non-linearly to preserve readability"). I suppose you are suggesting that we expand that a bit to say that UAs should determine (based on the factors you list and maybe others) if linear vs non-linear scaling would be best. If that's what you mean, that seems good, I'll put it in the draft.

As for this part
> Have you considered keeping _only_ `legacy` and `scale` as values

Yeah, we decided to keep only those values. But we'll obviously consider any future author requests for other scaling modes.

>   As you say in the explainer, the reason that `text-scale` has to be opt-in is potential breakage of _existing_ sites. However, new sites that are designed with a more adaptive approach to text sizing could be much more robust against whether the scale is linear or otherwise.
>   Have you considered explicitly making this choice up to the UA?

We hadn't considered that. (Unless I've forgotten and @JoshTumath remembers.) But it seems like a great space for UA innovation -- to analyze if the page would scale appropriately even in the absence of an explicit `text-scale`. That would be great. (Let me know if I've understood that correctly.)

> * Regarding the [question as to whether the font size keywords should also scale](https://github.com/w3c/csswg-drafts/issues/12475): we think this makes sense and should be included.

Agreed. To be clear other font size keywords will definitely scale. https://github.com/w3c/csswg-drafts/issues/12475 was about the further question of if they should scale linearly vs non-linearly.

> * We'd agree in expecting it to be very likely there would be significant breakage of existing site layouts if scaling were the default. But out of curiosity: have you done any testing to check on this?

Yes, but we unfortunately didn't document it. So, just now I did a test simulating default scaling at 1.7x and browsed around a bit. baltimoresun.com is a stereotypical example of breakage. The screenshot shows that the headlines and blurbs are cut off, which is bad enough, but much worse is that the user is even _unable to horizontally scroll_ to see the cropped text. We found many examples like this, unfortunately.

<img width="300" alt="scaled baltimoresun homepage" src="https://github.com/user-attachments/assets/125d334e-aaf3-44b5-882f-3699c29d6c6a" />


> Regarding the explainer:
> 
> * It's a well-written document, but it would be really helpful to have more illustrative examples - at least one "before and after" style example - of the effect of the proposal.

1000% agreed. Will fix.

> * Our read on the proposal is that it _would_ affect the base font size on desktop too (i.e. as well as mobile) - but [the table](https://github.com/w3c/csswg-drafts/blob/main/css-env-1/explainers/meta-text-scale.md#comparison-of-legacy-and-scale) makes this seem a little unclear - could you clarify?

Yes, it would affect base font size on desktop too. That's what we were trying to say with the "font-size: medium on desktop" row having "legacy: 16px × UA-level scale" vs "scale: 16px × OS-level scale × UA-level scale". (Suggestions for making that clearer are welcome. Perhaps we should explicitly say that font-size:medium is the base font size?)

Edit to add: though to be clear, this is a feature that we developed because of author feedback on mobile, and the mobile web will see the biggest usability improvements from this.

> Concerns regarding adoption:
> 
> From experience working with developers, the huge barrier this has to get over is that people need to opt in. It may be very hard to persuade busy development teams to opt in without some substantial resources to help them get started with such layouts. These may include:
> 
> * Guidance for authors in the spec.
> * [WAI Tutorials](https://www.w3.org/WAI/tutorials/) for authors to make more adaptive/robust/accessible layouts.
> * Making PRs on established projects such as
>   
>   * React Starter
>   * Svelte
>   * anything to do with bootstrapping projects
> 
> Getting this into the default `<head>` for projects will help achieve adoption - but it will also need some helpful resources for busy developers to ensure that the DX is still good.
> 
> You might want to contact [Docs CG](https://www.w3.org/community/docs-cg/) to see if they can assist you with writing the developer documentation.

Yes, agreed. We are planning a bit of a PR push once the feature has shipped and stabilized for a few milestones. I see that https://www.w3.org/WAI/tips/developing/#write-code-that-adapts-to-the-users-technology is a candidate for where some instruction might live. We also have a few ideas for chrome's developer tooling to make this easier on authors, one of which we already implemented (simulating changing the OS text scale factor.) I'd considered pushing vscode to get this meta tag in their default HTML template, similar to meta viewport, but attacking it at the framework level also seems prudent. Thanks for the suggestion.

> We look forward to hearing your thoughts on this.
> 
> Thanks to [@AutoSponge](https://github.com/AutoSponge) in APA WG for the benefit of his experience managing development.

**Comment by JoshTumath** [comment]
> > As you say in the explainer, the reason that `text-scale` has to be opt-in is potential breakage of _existing_ sites. However, new sites that are designed with a more adaptive approach to text sizing could be much more robust against whether the scale is linear or otherwise.
> > Have you considered explicitly making this choice up to the UA?
> 
> We hadn't considered that. (Unless I've forgotten and [@JoshTumath](https://github.com/JoshTumath) remembers.) But it seems like a great space for UA innovation -- to analyze if the page would scale appropriately even in the absence of an explicit `text-scale`. That would be great. (Let me know if I've understood that correctly.)

No we hadn't discussed that, but it sounds interesting. I'm sceptical about whether it's possible because of how unpredictable the existing text scaling is with `text-size-adjust: auto`; some areas of text grow while others don't and it's not clear to the author when it will happen. But if a UA is able to make that work, I'd be all for it!

**Comment by matatk** [comment]
Thank you @davidsgrogan and @JoshTumath for your replies, clarifications, and updates to the explainer, and spec. There's one thing on which I think I was not clear, and would like to check with you...

The current draft has the `legacy` and `scale` modes. The explainer mentions a possible `progressive` mode that could be added in future. But would it really be needed?

I think you've both answered the question "Could the UA try to decide if a legacy page could be scaled (without breakage, or burdensome scrolling)?" - I wasn't meaning to ask that question, but your answer is interesting, and encouraging.

The question I meant to ask is: "Given a page that opts into linear scaling via `scale`, wouldn't you expect that its layout would be more robust, and thus be able to cope with the non-linear scaling (described as `progressive` in the explainer) at any time, as determined by the UA, and user's preferences?"

An accessibility concern I have here is that I think the user ought to have control over what type of scaling they need (UAs could even offer control over the scaling factors, if they liked - I think I would find this useful). So even if other scaling modes are added in future, what I'm asking for is to allow the UA to always be able to make the final decision as to how to scale the fonts. (That aligns with the SHOULD in the spec that you highlighted.)

(Due to the breakage issues such as those you found, I wasn't meaning to include `legacy` pages in this, but if experimentation reveals UAs _could_ actually reliably determine that no two-axis scrolling would be needed for a given layout, it would seem to be fine to allow UAs to make the decision on `legacy` pages, too.)

Particularly regarding `scale` vs future additional modes: I hope that clears things up - how does it sound to you?

**Comment by davidsgrogan** [comment]
Thanks for the clarifications. I did indeed miss your point but get it now.

> The explainer mentions a possible `progressive` mode that could be added in future. But would it really be needed?

You're right; a 'progressive' mode available to authors is probably unneeded.

> what I'm asking for is to allow the UA to always be able to make the final decision as to how to scale the fonts

That makes sense. What we'll probably do in the future, pure speculation at this point, is make progressive-scaling the default when a site asks for `scale` and then consider a user-setting if users miss linear scaling.

**Comment by matatk** [comment]
I didn't close the loop on this, but per our discussion (thanks again for your review request, and discussion, @davidsgrogan, @JoshTumath) we are satisfied with this proposal, and am closing this issue. All the best with your remaining work on this.
